Automattic/jetpack
# Extra options to configure cache htconfig

As [pointed out by a forum user](https://wordpress.org/support/topic/allow-filtering-the-htaccess-entry-expiresbytype-text-html-for-expert-mode/#post-16273621), users can override their `Cache-Control` header by defining a `WPSC_CACHE_CONTROL_HEADER` header.

However, Super Cache [also writes out hard-coded ExpiresByType headers](https://github.com/Automattic/jetpack/blob/e616302a144db2f9737414a3006559dda73ed7ab/projects/plugins/super-cache/wp-cache.php#L3108) which may contradict a custom `Cache-Control` header.

This PR fixes the issue by updating the way headers and expiry rules are managed in WP Super Cache (while still respecting the pre-existing defines). Now users can override either the headers or mod_expiry rules using two new filters: 
- `wpsc_htaccess_mod_headers`, and
- `wpsc_htaccess_mod_expires`.

If the `Cache-Control` value has a `max-age` setting (whether from the default, define, or filter) it automatically tries to set an appropriate `ExpiresByType` rule which matches. For more complex cases, users can define their own mod_expires rules.

**Examples**
```
add_filter( 'wpsc_htaccess_mod_headers', function( $headers ) {
    $headers['Cache-Control'] = 'max-age=31530, immutable';
    $headers['Accept-Ranges'] = 'bytes';

    return $headers;
}, 1 );

add_filter( 'wpsc_htaccess_mod_expires', function( $expiry_rules ) {
    $expiry_rules = array_filter( $expiry_rules, function( $rule ) {
        return strpos( $rule, 'ExpiresByType' ) === false;
    } );

    $expiry_rules[] = 'ExpiresByType text/html "access plus 1 year"';

    return $expiry_rules;
}, 1 );
```

Fixes #27851

#### Changes proposed in this Pull Request:
* Add `wpsc_htaccess_mod_expires` and `wpsc_htaccess_mod_headers` to allow users to customize their htaccess file.

#### Other information:

#### Jetpack product discussion
n/a

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
* Enable advanced mode
* Try setting a `Cache-Control` header via the `WPSC_CACHE_CONTROL_HEADER` define, and ensure it is reflected in the `.htaccess` file in `wp-content/cache`
* Try adding a custom header and/or modifying the existing headers in the above htaccess file by defining a `wpsc_htaccess_mod_headers` filter.
* Try the same for the `mod_expires` rules using the `wpsc_htaccess_mod_expires` filter.
